File: leetcode-js/003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    if (s.length === 0) {
        return 0;
    }

    const map = new Map();
    let max = 0;
    let j = 0;

    for (const [i, char] of s.split('').entries()) {
        if (map.has(char)) {
            j = Math.max(j, map.get(char) + 1);
        }
        map.set(char, i);
        max = Math.max(max, i - j + 1);
    }

    return max;
};

console.log(lengthOfLongestSubstring("pwwkew"));

File: leetcode-java/003-longest-substring-without-repeating-characters.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() == 0) {
            return 0;
        }
        int[] map = new int[256];
        int max = 0;
        for (int i = 0, j = 0; i < s.length(); ++i) {
            if (map[s.charAt(i) - ' '] != 0) {
                j = Math.max(j, map[s.charAt(i) - ' ']);
            }
            map[s.charAt(i) - ' '] = i + 1;
            max = Math.max(max, i - j + 1);
        }
        return max;
    }
}
